feat: Add client credentials guide for NGINXaaS Google (#2052)
nginx/documentation: content/nginxaas-google/quickstart/security-controls/client-credentials.md

---

> title: "Programmatic Authentication With Client Credentials"

Reviewer: ```suggestion
title: "Programmatic authentication with client credentials"
```

Use sentence case on the title and all headings

description: "Learn how to set up OAuth2 client credentials for programmatic access to the NGINXaaS API."
weight: 100
toc: true
url: /nginxaas/google/quickstart/security-controls/client-credentials/
f5-product: NGOOGL
f5-content-type: how-to
f5-keywords: "client credentials, OAuth2, programmatic authentication, API"
---

## Overview

> This guide explains how to create and use client credentials to manage some NGINXaaS for Google Cloud APIs programmatically. This allows machine users such as Terraform, CI/CD pipelines, or other automation tools to authenticate and interact with the NGINXaaS API securely.

Reviewer: ```suggestion
This guide explains how to create and use client credentials to manage some NGINXaaS for Google Cloud APIs programmatically. Client credentials allow machine users such as Terraform, CI/CD pipelines, or other automation tools to authenticate and interact with the NGINXaaS API securely.
```
Avoid vague phrasing

Client credentials are scoped to an organization and enable programmatic access to manage Deployments, Configurations, and Certificates APIs.

## Key features

- **Secure programmatic access**: Enables programmatic access for machine clients with tokens minted using client credentials
- **Configurable Expiration controls**: Configurable expiration with recommended best practices

> - **Limited scope**: Client credentials can only manage Deployments, Configs, and Certificates

Reviewer: ```suggestion
You set an expiration when you create the credentials, up to a maximum of one year, with six months recommended.
```

How-tos usually don't have key-features (these are reserved for marketing materials). The relevant content in this section can be added to the Overview section with just once sentence.

Author: removed key features from doc

Reviewer 2: I still see the entire section so I am not sure if the change is gone. I think we can capture this detail well enough in the overview section like Jon mentioned!!

Author: its not there for me. are you seeing the latest version?

> ## Prerequisites

Reviewer: ```suggestion
## Before you begin
```

- You must be logged into NGINXaaS Console [https://console.nginxaas.net/](https://console.nginxaas.net/)

> ## Creating client credentials

Reviewer: ```suggestion
## Create client credentials
```

Follow these steps to create a new client credential through the NGINXaaS console:

1. Log in to [NGINXaaS Console](https://console.nginxaas.net/) using your Google credentials.
2. Select **Settings** > **Client Credentials** from the left navigation menu.
3. Select **+ Add Client** to create a new credential.
4. Enter a unique name for your client in the text field that appears.
5. Choose an expiration date (maximum 1 year from today; 6 months is recommended). This setting is immutable after creation.
6. Select **Create**. A popup window displays your client secret.

{{< call-out class="warning" title="Important: Store Your Client Secret Securely" >}}
The client secret appears only once. Save it immediately in a secure location, such as a password manager or secrets vault. If you lose the secret, you must delete this credential and create a new one.
{{< /call-out >}}

Your client credentials can access the following APIs programmatically:

- Certificates API
- Configs API
- Deployments API

### Client limits

Organizations are limited to a maximum of 10 client credentials. To request an increase to this limit, contact the NGINX Support team.

> ## Retrieving client information

Reviewer: ```suggestion
## Retrieve client information
```

Follow these steps to view your client credentials:

1. Log in to [NGINXaaS Console](https://console.nginxaas.net/).
2. Select **Settings** > **Client Credentials** from the left navigation menu.

> 3. The **Client Credentials** page displays all available client credentials for your organization. The table shows the following metadata for each credential:

Reviewer: ```suggestion
3. You can see all the available client credentials for your organization in the **Client Credentials** page. The table shows the following metadata for each credential:
```

Steps needs to be in action form

1. Log in to [NGINXaaS Console](https://console.nginxaas.net/).
2. Select **Settings** > **Client Credentials** from the left navigation menu.
3. Locate the credential to delete in the table.
4. Select the three vertical dots icon (⋮) at the end of the row.
5. Select **Delete** from the menu.

After deletion:

- The client can no longer request new access tokens using this credential
- Existing access tokens continue to work until they expire (1 hour from issuance)
- The client ID and secret become permanently invalid

{{< call-out class="note" >}}
Expired credentials are not automatically removed. You must manually delete credentials that are past their expiration date.
{{< /call-out >}}

## Using client credentials for authentication

### Exchange credentials for an access token

Use the client credentials to obtain an access token from the token endpoint.

**Endpoint**: `POST https://us.api.nginxaas.net/api/v1/marketplace/auth/token`

**Example using cURL**:

```bash
curl -X POST "https://us.api.nginxaas.net/api/v1/marketplace/auth/token" \
  -H "Content-Type: application/json" \
  -d '{
    "client_id": "client_abc123",
    "client_secret": "superSecret",

>     "grant_type": "client_credentials"

Reviewer: if these are placeholders they need to be in a <ABC_XYZ> form

Author: updated everywhere

>     "grant_type": "client_credentials"

Reviewer 2: realllllly 😆

  }'
```

**Response**:

```json
{

>   "access_token": "<access_token>",

Reviewer: ```suggestion
  "access_token": "<ACCESS_TOKEN>",
```

  "token_type": "Bearer",
  "expires_in": 3600
}
```

### Use the access token

Include the access token in the Authorization header when making API requests.

**Example**:

```bash
curl -X GET "https://us.api.nginxaas.net/api/v1/deployments" \

>   -H "Authorization: Bearer <access_token>"

Reviewer: ```suggestion
  -H "Authorization: Bearer <ACCESS_TOKEN>"
```

we have placeholder rules

>   -H "Authorization: Bearer eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCJ9..."

Reviewer 3: I just got a security alert on this

<img width="1181" height="254" alt="Image" src="https://github.com/user-attachments/assets/88f357ad-5e0e-4474-86d8-acaac0cabeb7" />

Author: i replaced with a variable name. it was just garbage token

Reviewer 2: Just as a general practice, try not to commit such things in docs. It can be `${JWT}` or `Authorization: Bearer this.Is.MyJWT` or something.

Author: yep removed it

```

### Access token validity

- **Duration**: Access tokens are valid for 1 hour from issuance
- **Reauthentication**: When a token expires, request a new one using the same client credentials, if the client credentials are still active
- **Scope**: Tokens are also scoped to the organization associated with the client credentials

## Security best practices

> - **Use separate credentials for different environments**: Create distinct clients for various scopes to keep things logically separate

Reviewer: ```suggestion
- **Use separate credentials for different environments**: Create distinct clients for various scopes to keep environments logically separate
```
avoid vague phrasing

Reviewer: feel free to replace my suggestion with another specific term

Author: environments seems right

- **Store secrets securely**: Never commit client secrets to version control
- **Delete unused credentials**: Remove clients that are no longer needed
- **Follow the recommended expiration**: Unless you have a specific reason, use the recommended 6-month expiration

> ## Error handling

Reviewer: ```suggestion
## Troubleshooting
```

- Verify your client ID and secret are correct
- Check if the client has expired
- Create new credentials if necessary

**Example error response**:

```json
{
  "error": "unauthorized",
  "error_description": "token is expired/invalid"
}
```

### Forbidden APIs

Client credentials can only access the Certificates, Configs, and Deployments APIs. Attempting to access any other API with client credentials returns a `403 Forbidden` response. In this case:

- Verify you are using the correct API endpoint
- Ensure the API is one of the supported resources (Deployments, Configs, or Certificates)

**Example error response**:

```json
{
  "error": "forbidden",
  "error_description": "Client credentials do not have permission to access this resource"
}
```

## Quick reference

| Property | Value |
|----------|-------|
| Client secret visibility | Only shown once during creation |
| Default expiration | 6 months (recommended) |
| Maximum expiration | 1 year |
| Client limit per organization | 10 clients (contact NGINX Support to increase) |
| Access token validity | 1 hour |
| Supported resources | Deployments, Configs, Certificates |

> | Token endpoint | `https://{region}.api.nginxaas.net/api/v1/marketplace/auth/token` |

Reviewer: there should be a `## References` section at the end of the doc, linking to related NGINX docs and external sources as necessary

Author: have added this as well

> | Token endpoint | `https://{region}.api.nginxaas.net/api/v1/marketplace/auth/token` |

Reviewer: ```suggestion
| Token endpoint | `https://<REGION>.api.nginxaas.net/api/v1/marketplace/auth/token` |
```
rules for placeholders.

Reviewer 2: I have not looked at the latest changes but the correct term to use here is `<GEO>`. Please change this label to leverage Geos instead.

---
title: "Programmatic authentication with client credentials"
description: "Learn how to set up OAuth2 client credentials for programmatic access to the NGINXaaS API."
weight: 100
toc: true
url: /nginxaas/google/quickstart/security-controls/client-credentials/
f5-product: NGOOGL
f5-content-type: how-to
f5-keywords: "client credentials, OAuth2, programmatic authentication, API"
f5-summary: >

>     Use this guide to create OAuth2 client credentials in the NGINXaaS for Google Cloud Console and exchange

Reviewer 2: ```suggestion
    Use this guide to create OAuth2 client credentials in the NGINXaaS Console and exchange
```

>     them for an access token that authenticates programmatic requests to the NGINXaaS API.

Reviewer 2: ```suggestion
    them for an access token that authenticates requests to the NGINXaaS API.
```

>     Client credentials let machine users such as Terraform, CI/CD pipelines, and other automation tools

Reviewer 2: let's stick to automation toolchains and not leverage `machine users`. It might end up being slightly confusing for the end user.

Maybe we rephrase it along the lines of `Client credentials let you manage your deployments, configurations, certificates in automation flows such as CICD.`

Also please redact Terraform for now. We don't have support for it yet and it will be confusing if someone tries looking for our provider.

    authenticate without a human login.

>     Credentials are scoped to a single organization and can only manage the Deployments, Configurations, and

Reviewer 2: what a user is allowed to manage:

this is an important detail worth highlighting in a note.

    Certificates APIs.
f5-audience: operator
---

## Overview

> This guide explains how to create and use client credentials to manage some NGINXaaS for Google Cloud APIs programmatically. Client credentials allow machine users such as Terraform, CI/CD pipelines, or other automation tools to authenticate and interact with the NGINXaaS API securely.

Reviewer 2: This is a generic comment and should be addressed more broadly.

Programmatic auth is applicable regardless of the target cloud as it is a feature of NGINXaaS (not Google or cloud specific). You should reword these changes accordingly.

Reviewer 2: also this comment around machine users: https://github.com/nginx/documentation/pull/2052/changes#r3431828589

Client credentials are scoped to an organization and enable programmatic access to manage Deployments, Configurations, and Certificates APIs.

## Key features

- **Secure programmatic access**: Enables programmatic access for machine clients with tokens minted using client credentials
- **Configurable expiration controls**: Configurable expiration with recommended best practices
- **Limited scope**: Client credentials can only manage Deployments, Configs, and Certificates

## Before you begin

- You must be logged in to the [NGINXaaS Console](https://console.nginxaas.net/).

## Create client credentials

Follow these steps to create a new client credential through the NGINXaaS console:

> 1. Log in to [NGINXaaS Console](https://console.nginxaas.net/) using your Google credentials.

Reviewer 2: ```suggestion
1. Log in to [NGINXaaS Console](https://console.nginxaas.net/).
```

You may want to keep it to that. if we add additional auth mechanisms then we are dealing with a missed update here 🤷

2. Select **Settings** > **Client Credentials** from the left navigation menu.

> 3. Select **+ Add Client** to create a new credential.

Reviewer 2: @JTorreG can correct me but I recall not specifically creating a numeric list `1., 2., 3.....1` and instead had to just repeat `1.` across and markdown took care of autonumbering or something. I forget.

There are examples of this all over the docs and we should follow the same here.

Author: this is inline with the format expected.

> 4. Enter a unique name for your client in the text field that appears.

Reviewer 2: do we enforce the uniqueness on it?

Author: i assumed auth0 would

5. Choose an expiration date (maximum 1 year from today; 6 months is recommended). This setting is immutable after creation.

> 6. Select **Create**. A popup window displays your client secret.

Reviewer 2: nit: pop up is a next step 😆

{{< call-out class="warning" title="Important: Store Your Client Secret Securely" >}}

> The client secret appears only once. Save it immediately in a secure location, such as a password manager or secrets vault. If you lose the secret, you must delete this credential and create a new one.

Reviewer 2: two cents:

I usually prefer not strongly recommending to a customer around deletions as often.

I'd usually phrase it calling out that if they do not save, they won't be able to retrieve the credentials later on, thus making the client unusable.

If you cannot use it since you lost the secret, deletion is implied here 🤷 .

Author: removing it

{{< /call-out >}}

> Your client credentials can access the following APIs programmatically:

Reviewer 2: this is called out in several places. let's keep this list in one spot.

If you feel it is necessary to repeat the section for a certain case, then let's create an `includes` for it and reference that instead.

I think calling this out once in a highlighted way should do it?

- Certificates API
- Configs API

> - Deployments API

Reviewer 2: also.

I have a larger concern: If I am reading this with no context, the immediate next question is: What is the deployments API. where do I find docs on it?

Author: i added that in the references section at the end of this page

Reviewer 2: does it make sense to link them here as well? I think we should.

Author: added

### Client limits

> Organizations are limited to a maximum of 10 client credentials. To request an increase to this limit, contact the NGINX Support team.

Reviewer 2: can we please href this to the support page?

## Retrieve client information

> Follow these steps to view your client credentials:

Reviewer 2: ```suggestion
Follow these steps to view information about your clients:
```
?

After deletion:

- The client can no longer request new access tokens using this credential
- Existing access tokens continue to work until they expire (1 hour from issuance)
- The client ID and secret become permanently invalid

{{< call-out class="note" >}}
Expired credentials are not automatically removed. You must manually delete credentials that are past their expiration date.
{{< /call-out >}}

## Using client credentials for authentication

### Exchange credentials for an access token

Use the client credentials to obtain an access token from the token endpoint.

**Endpoint**: `POST https://{region}.api.nginxaas.net/api/v1/marketplace/auth/token`

**Example using cURL**:

```bash
curl -X POST "https://{region}.api.nginxaas.net/api/v1/marketplace/auth/token" \
  -H "Content-Type: application/json" \
  -d '{
    "client_id": "<client_id>",
    "client_secret": "<client_secret>",
    "grant_type": "client_credentials"
  }'
```

**Response**:

```json
{
  "access_token": "<ACCESS_TOKEN>",
  "token_type": "Bearer",
  "expires_in": 3600
}
```

### Use the access token

Include the access token in the Authorization header when making API requests.

**Example**:

```bash

> curl -X GET "https://{region}.api.nginxaas.net/api/v1/deployments" \

Reviewer 2: ```suggestion
curl -X GET "https://{geo}.api.nginxaas.net/api/v1/deployments" \
```

  -H "Authorization: Bearer <ACCESS_TOKEN>"
```

### Access token validity

- **Duration**: Access tokens are valid for 1 hour from issuance
- **Reauthentication**: When a token expires, request a new one using the same client credentials, if the client credentials are still active
- **Scope**: Tokens are also scoped to the organization associated with the client credentials

## Security best practices

> - **Use separate credentials for different environments**: Create distinct clients for various scopes to keep environments logically separate

Reviewer 2: this is where we are delving into use-cases. This is just general feedback on it:

- I think having a best practices section is good 👍 
- Recommended expiration is just a sane default if a customer does not want to think. If we want to offer a recommendation, we should call out not using credentials with a long expiration.
- Storage of creds: There are other best practices as well other than not committing it in source code but try going the other way, i.e, we could state that the user should consider storing credentials in a secure place of their choice and defer to them as to what that place is.
- `Use separate credentials for different environments`: This is generally true but we have an assumption around user envs. I might choose to segregate credentials per deployment region🤷 . If we want to make a recommendation, maybe state something that keeps this vague while offering guidance: `consider leveraging different clients for different use-cases..` << this is going to get confusing if not worded properly.

Author: have updated this

- Verify your client ID and secret are correct
- Check if the client has expired
- Create new credentials if necessary

**Example error response**:

```json
{
  "error": "unauthorized",
  "error_description": "token is expired/invalid"
}
```

### Forbidden APIs

Client credentials can only access the Certificates, Configs, and Deployments APIs. Attempting to access any other API with client credentials returns a `403 Forbidden` response. In this case:

- Verify you are using the correct API endpoint
- Ensure the API is one of the supported resources (Deployments, Configs, or Certificates)

**Example error response**:

```json
{
  "error": "forbidden",
  "error_description": "Client credentials do not have permission to access this resource"
}
```

## Quick reference

| Property | Value |
|----------|-------|
| Client secret visibility | Only shown once during creation |
| Default expiration | 6 months (recommended) |
| Maximum expiration | 1 year |
| Client limit per organization | 10 clients (contact NGINX Support to increase) |
| Access token validity | 1 hour |
| Supported resources | Deployments, Configs, Certificates |

> | Token endpoint | `https://<REGION>.api.nginxaas.net/api/v1/marketplace/auth/token` |

Reviewer 4: The NGINXaaS terminology is Geographic Controller https://github.com/nginx/documentation/blob/main/content/nginxaas-google/overview.md#geographical-controllers

region is a cloud-provider term for where a deployment may be deployed.

```suggestion
| Token endpoint | `https://<geo>.api.nginxaas.net/api/v1/marketplace/auth/token` |
```

Also, I think we want `/api/v1/auth/token` ?

Reviewer 2: yeah we do except that the change for it is in flight.

If docs land first, then we can update this endpoint as a follow up. I'd prefer that we display accurate information here though. If that's the current endpoint that works, then let's keep that until we make the change to remove `marketplace`. otherwise, users will get confused as they will get a 404.

Author: have changed to geo and also added the /auth/token end pointthat corresponds to what is in production right now
